### PR TITLE
Fixes #2358.  Added to the reducer the ability to remove a function t…

### PR DESCRIPTION
**Pull request by afd**
…hat is not directly called.  Factored out some code from the optimizer to help with this.

### CODE REVIEW

File: source/reduce/remove_function_reduction_opportunity_finder.cpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+namespace spvtools {
+namespace reduce {
+
+std::vector<std::unique_ptr<ReductionOpportunity>>
+RemoveFunctionReductionOpportunityFinder::GetAvailableOpportunities(
+    opt::IRContext* context) const {
+  std::vector<std::unique_ptr<ReductionOpportunity>> result;
+  // Consider each function.
+  for (auto& function : *context->module()) {
+    if (context->get_def_use_mgr()->NumUses(function.result_id()) > 0) {
```
[the remark this answers is not in the document]
Comment (alan-baker) [reply source/reduce/remove_function_reduction_opportunity_finder.cpp]: The recursion restriction is actually added by Vulkan and is not core to SPIR-V. Not sure if this should only handle Vulkan cases.

File: source/opt/eliminate_dead_functions_util.h [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+#include "source/opt/ir_context.h"
+
+namespace spvtools {
+namespace opt {
+
+// Provides functionality for eliminating functions that are not needed, for use
+// by various analyses and passes.
+namespace eliminatedeadfunctionsutil {
+
+// Remove all of the instruction in the function body
```
[the remark this answers is not in the document]
Comment (afd) [reply source/opt/eliminate_dead_functions_util.h]: Done (changed comment too.)

File: source/reduce/remove_function_reduction_opportunity.cpp [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+bool RemoveFunctionReductionOpportunity::PreconditionHolds() {
+  // Removing one function cannot influence whether another function can be
+  // removed.
+  return true;
+}
+
+void RemoveFunctionReductionOpportunity::Apply() {
+  opt::eliminatedeadfunctionsutil::EliminateFunctionInstructions(context_,
+                                                                 function_);
+  for (opt::Module::iterator function_it = context_->module()->begin();
```
[the remark this answers is not in the document]
Comment (afd) [reply source/reduce/remove_function_reduction_opportunity.cpp]: Done.

File: source/reduce/remove_function_reduction_opportunity_finder.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (afd) [reply source/reduce/remove_function_reduction_opportunity_finder.cpp]: In general, where removal of Xs is blocked due to Ys not being removed, we should gradually file issues for, and then add features to, remove Ys.  So while we definitely need to think more on removal of decorations, that should not affect this pass.

The issue of recursion is potentially interesting and I guess would be a problem if other passes are not able to break recursive cycles (by removing OpCall instructions).

I suggest we revisit that in future in response to modules that don't reduce well due to this, if we find them.

**Comment by afd** [comment]
Clang format checks are to do with changes elsewhere.

File: source/reduce/remove_function_reduction_opportunity_finder.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (s-perron) [reply source/reduce/remove_function_reduction_opportunity_finder.cpp]: I agree.  Keep it simply for now, and see how all of the reduction opportunities work together.  Improve as needed.

**Pull request merged**